### PR TITLE
ROSAENG-454 Add close-on-release feature to delete unused AWS accounts

### PR DESCRIPTION
**Pull request by dakotalongRH**
Previously, when an AccountClaim was deleted, AAO would clean up AWS resources and return the Account to the pool for reuse. Over time, this led to thousands of dormant osd-creds-* accounts accumulating in AWS Organizations.

This change adds a new feature flag that, when enabled, closes the AWS account via the Organizations CloseAccount API instead of reusing it.

Changes:
- Add CloseAccount and DescribeAccount methods to AWS client interface
- Add feature.close_on_release ConfigMap flag (default: disabled)
- Add close_account.dry_run ConfigMap flag (default: enabled for safety)
- Implement closeAndDeleteAccount() in accountclaim controller
- Add exponential backoff for AWS rate limits (1h-24h, stored in Account CR annotations)
- Add aws_account_operator_accounts_closed_total metric
- Skip closure for FedRAMP, CCS/BYOC, and STS accounts

The feature uses safe defaults:
- Disabled by default (no behavior change)
- Dry-run enabled by default (logs only, no API calls)
- Falls back to reuse if closure fails

AWS rate limit handling:
- AWS allows closing max(250, 20% of accounts) per rolling 30-day window
- On CLOSE_ACCOUNT_QUOTA_EXCEEDED, sets exponential backoff via annotations
- Automatically retries as quota frees up

# What is being added?
_Is this a fix for a bug?  What's the bug?  Is this a new feature? Please describe it. Is this just a small typo fix?  That's fine too!_

## Checklist before requesting review

- [ ] I have tested this locally
- [X] I have included unit tests
- [X] I have updated any corresponding documentation

## Steps To Manually Test
_Please provide us steps to reproduce the scenarios needed to test this.  If an integration test is provided, let us know how to run it. If not, enumerate the steps to validate this does what it's supposed to._
Step 1: Deploy with Dry-Run (Safe)
# Edit the ConfigMap
oc edit cm aws-account-operator-configmap -n aws-account-operator
Add:

data:
  feature.close_on_release: "true"
  close_account.dry_run: "true"
Step 2: Delete an AccountClaim
# Find a test AccountClaim
oc get accountclaims -A
# Delete it
oc delete accountclaim <name> -n <namespace>
Step 3: Check Logs
oc logs -n aws-account-operator deployment/aws-account-operator -f | grep -E "DRY-RUN|close"
Expected output:

INFO DRY-RUN: Would close account  accountID=123456789012 accountCR=osd-creds-mgmt-abc123
Step 4: Enable Real Closure (When Ready)
oc patch cm aws-account-operator-configmap -n aws-account-operator \
  --type merge -p '{"data":{"close_account.dry_run":"false"}}'
Step 5: Monitor
# Watch for closures
oc logs -n aws-account-operator deployment/aws-account-operator -f | grep -E "Successfully closed|rate limit|CloseAccount"
# Check metrics
curl -s <metrics-endpoint> | grep accounts_closed
Step 6: Verify in AWS Console
After a real closure, check AWS Organizations console:

Account should show status PENDING_CLOSURE or SUSPENDED

## Release Notes

* **New Features**
  * Added AWS account closure capability when released (controlled by "close-on-release" feature flag).
  * Added dry-run mode support for testing closure operations without AWS impact.
  * Implemented rate-limiting and exponential backoff handling for account closure quota management.
  * New metrics to track the total number of closed accounts.

* **Tests**
  * Added comprehensive unit tests for account closure utilities and feature configuration.

### CODE REVIEW

File: controllers/accountclaim/reuse.go [hunk 1]
```diff
@@ -200,6 +226,101 @@ func (r *AccountClaimReconciler) resetAccountSpecStatus(reqLogger logr.Logger, r
...
+	return nil
+}
+
+// isCloseAccountRateLimitError checks if the error is a rate limit error from CloseAccount
+func isCloseAccountRateLimitError(err error) bool {
+	var constraintErr *orgtypes.ConstraintViolationException
+	if errors.As(err, &constraintErr) {
+		if constraintErr.Message != nil &&
+			strings.Contains(*constraintErr.Message, "CLOSE_ACCOUNT_QUOTA_EXCEEDED") {
+			return true
```
[the remark this answers is not in the document]
Comment (dakotalongRH) [reply controllers/accountclaim/reuse.go:339]: done

**Comment by BATMAN-JD** [comment]
## Code Review: Add close-on-release feature

### Overview
This PR introduces a well-designed feature to close AWS accounts permanently when AccountClaims are deleted. The implementation has excellent safety defaults, comprehensive error handling, and thorough test coverage.

---

### ✅ Strengths

1. **Excellent Safety Design**
   - Multiple safety layers: feature disabled by default, dry-run enabled by default, graceful fallback to reuse
   - Protected accounts: Correctly skips FedRAMP accounts
   - Already-closed handling: Gracefully handles `AccountAlreadyClosedException`

2. **Robust Rate Limit Handling**
   - Exponential backoff: 1h → 2h → 4h → 24h with sensible caps
   - Persistent state stored in CR annotations
   - Specific error detection for both AWS quota types

3. **Comprehensive Test Coverage**
   - 353 lines of tests covering edge cases
   - Clear test names and descriptions

4. **Good Observability**
   - New Prometheus metric: `aws_account_operator_accounts_closed_total`
   - Detailed logging at key decision points

---

### 🔍 Critical Issues

#### 1. **Missing CCS/BYOC/STS Check in closeAndDeleteAccount** ⚠️

**Location**: `controllers/accountclaim/reuse.go:165`

The PR description states: *"Skip closure for FedRAMP, CCS/BYOC, and STS accounts"*

However, the close-on-release check only skips FedRAMP:
```go
if utils.IsCloseOnReleaseEnabled(configMap) && !config.IsFedramp() {
```

**Problem**: CCS/BYOC/STS accounts could be incorrectly processed if close-on-release is enabled globally.

**Suggested Fix**:
```go
if utils.IsCloseOnReleaseEnabled(configMap) && 
   !config.IsFedramp() &&
   !reusedAccount.IsBYOC() &&
   !reusedAccount.Spec.BYOC &&
   !reusedAccount.Spec.ManualSTSMode {
    closeErr := r.closeAndDeleteAccount(reqLogger, reusedAccount, awsClient, configMap)
    // ...
}
```

#### 2. **Race Condition: Annotation Update Without Error Check**

**Location**: `controllers/accountclaim/reuse.go:276-281`

```go
account.Annotations = utils.SetCloseAccountRateLimited(account.Annotations)
if updateErr := r.Update(context.TODO(), account); updateErr != nil {
    reqLogger.Error(updateErr, "Failed to update account annotations for rate limit")
}
retryAfter := utils.GetCloseAccountRetryAfter(account.Annotations)
```

**Problem**: If `Update()` fails due to conflict, the in-memory annotations may not match cluster state, leading to incorrect retry calculations.

**Suggested Fix**:
```go
account.Annotations = utils.SetCloseAccountRateLimited(account.Annotations)
if updateErr := r.Update(context.TODO(), account); updateErr != nil {
    reqLogger.Error(updateErr, "Failed to update account annotations for rate limit")
    return fmt.Errorf("failed to persist rate limit state: %w", updateErr)
}
// Now safe to read from persisted state
retryAfter := utils.GetCloseAccountRetryAfter(account.Annotations)
```

---

### 🔧 Moderate Issues

#### 3. **Unclear Dry-Run Error Handling**

**Location**: `controllers/accountclaim/reuse.go:247-252`

Dry-run returns an error which triggers the fallback-to-reuse path that logs it as a failure. This makes dry-run appear as an **error** in logs when it's expected behavior.

**Suggested Fix** - Use a sentinel error:
```go
var ErrDryRunMode = errors.New("dry-run mode enabled")

// In finalizeAccountClaim:
closeErr := r.closeAndDeleteAccount(reqLogger, reusedAccount, awsClient, configMap)
if closeErr != nil {
    if errors.Is(closeErr, ErrDryRunMode) {
        reqLogger.Info("Dry-run mode active, falling back to reuse")
    } else {
        reqLogger.Error(closeErr, "Failed to close account, falling back to reuse", ...)
    }
}
```

#### 4. **Potential Context Leakage**

**Location**: `controllers/accountclaim/reuse.go:257`

Using `context.Background()` instead of inheriting the reconciler's context means cancellation signals won't propagate and operator shutdown may not cleanly abort in-flight calls.

**Suggested Fix**: Accept context as parameter:
```go
func (r *AccountClaimReconciler) closeAndDeleteAccount(
    ctx context.Context,  // Add this parameter
    reqLogger logr.Logger,
    // ... other params
) error {
    callCtx, cancel := context.WithTimeout(ctx, 30*time.Second)
    defer cancel()
    // ...
}
```

---

### 📋 Minor Issues

5. **Inconsistent Context Usage**: Mix of `context.Background()`, `context.TODO()` throughout
6. **Magic Number**: 30 second timeout should be a constant
7. **Dockerfile Version Bump**: Is the base image update related to this feature?
8. **Missing Documentation**: ConfigMap keys not documented in README
9. **Missing Integration Test**: No test for full close-on-release flow with mocked AWS client

---

### 📊 Performance & Security

**Performance**: ✅ Efficient design with good rate limiting

**Security**: 
- ✅ Payer account protection exists
- ✅ FedRAMP exclusion prevents compliance issues
- ❓ Verify operator IAM permissions include `CloseAccount` API
- 📝 Document 90-day `PENDING_CLOSURE` recovery period

---

### 🎯 Recommendations

**Must Fix** (blocking):
1. ❌ Add CCS/BYOC/STS account type checks before closing
2. ❌ Fix annotation update error handling

**Should Fix** (before merge):
3. ⚠️ Distinguish dry-run from actual errors in logging
4. ⚠️ Use inherited context instead of `Background()`/`TODO()`
5. ⚠️ Add integration test for full flow

**Nice to Have** (post-merge):
6. 📝 Document ConfigMap flags
7. 📊 Add metric labels for close status
8. 🔧 Add jitter to exponential backoff

---

### 📝 Summary

**Overall**: High-quality work with thoughtful safety mechanisms and excellent testing. The exponential backoff implementation is particularly well-designed for AWS's quota model.

**Recommendation**: **Request changes** to address the two critical issues (account type checks and annotation race condition), then approve after fixes.

**Risk**: Medium-Low (with fixes). The multi-layer safety approach provides good protection, but the account type filtering gap should be closed before production use.

Great work! 🎉

**Comment by BATMAN-JD** [comment]
I don't believe integrations test failed because of you. I am so sick of these things. I believe this is PROW again. I will retry them and see if it fails to start pods again...

**Comment by BATMAN-JD** [comment]
Multiple issues different with the PROW env yesterday. Was different each run. Going to try again today, and if it doesn't work we may just ignore the integration tests for this PR, I don't see anything that is breaking them and they are technically not needed for PR to merge

**Comment by BATMAN-JD** [comment]
The integration test failures on this PR are not caused by the code changes. This is a CI cluster infrastructure problem.

Evidence
Build Phase: ✅ Successful


Image ci-op-3k1v3wq4/pipeline:aws-account-operator created
digest=sha256:2acc492b6f5920401cb3d3a359313112f9745990d68e5171016e9582e3dd0235
Deployment Phase: ❌ Failed - Kubelet TLS Error


Pod status: aws-account-operator-56cd49db9c-xt642   0/1     Pending

Pod events:
110s   Normal   Scheduled   pod/aws-account-operator   Successfully assigned... to ip-10-0-11-49.ec2.internal
111s   Normal   SuccessfulCreate   replicaset/aws-account-operator   Created pod: aws-account-operator

Pod logs (if any):
Error from server: Get "https://10.0.11.49:10250/containerLogs/...": remote error: tls: internal error
No pod logs available
Root Cause
The pod is scheduled successfully but never starts because the kubelet on the assigned node cannot communicate via TLS. The container runtime never:

Pulls the image
Creates the container
Starts the container
This is why the pod remains stuck in Pending status after 120s.

Please just ignore these for now. I saw these running about two days ago on another pr just fine. These problems are not invoked by your changes. Not blocking PR to ignore.

**Comment by BATMAN-JD** [comment]
Integration tests randomly worked on #980 going to try and run again here and hopefully we are lucky between transient errors in PROW

**Comment by dakotalongRH** [comment]
Dry-run error handling - Added ErrDryRunMode sentinel error. Caller now uses errors.Is() to distinguish dry-run from actual failures and logs at Info level instead of Error.

Context propagation - closeAndDeleteAccount() now accepts a context parameter and derives timeouts from it instead of using context.Background(). This ensures cancellation signals propagate correctly during operator shutdown.

Annotation update error handling - Now returns the error if Update() fails, ensuring rate limit state is persisted before proceeding.

Added closeAccountTimeout constant for the 30-second timeout.

For the CCS/BYOC/STS accounts comment:
These account types are already excluded - they return early in finalizeAccountClaim() before close-on-release code is reached:

STS (lines 60-68): Checks reusedAccount.Spec.ManualSTSMode → returns
BYOC/CCS (lines 113-128): Checks reusedAccount.IsBYOC() → returns
FedRAMP (line 165): Explicitly checked in the close-on-release condition
No additional guards needed since the close-on-release logic is only reachable for non-BYOC, non-STS, non-FedRAMP pool accounts.

**Comment by eth1030** [comment]
I think everything looks good but you missed https://github.com/openshift/aws-account-operator/pull/978#discussion_r3185938223  so the current code is putting rate-limited accounts back in the pool. instead we would want to exit early before calling `resetAccountSpecStatus`

**Comment by dakotalongRH** [comment]
> I think everything looks good but you missed [#978 (comment)](https://github.com/openshift/aws-account-operator/pull/978#discussion_r3185938223) so the current code is putting rate-limited accounts back in the pool. instead we would want to exit early before calling `resetAccountSpecStatus`

fixed in the latest push. Rate-limited accounts now return early with ErrRateLimited before reaching resetAccountSpecStatus, so they stay claimed and the finalizer will retry based on the backoff annotations. Only dry-run and other non-retryable errors fall through to reuse.

**Comment by BATMAN-JD** [comment]
same transient error in PROW that happened at the end of last week too. Ignore integration-test for now. Lame because they have worked all week for all other PR's, but I confirmed this with another test branch I have
